presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetReader.java

```diff
@@ -151,102 +155,203 @@ private boolean advanceToNextRowGroup()
+        List<TypeSignatureParameter> fields = type.getTypeSignature().getParameters();
+        Block[] blocks = new Block[fields.size()];
+        IntList fieldDefinitionLevels = null;
+        IntList fieldRepetitionLevels = null;
+        for (int i = 0; i < fields.size(); i++) {
+            fieldDefinitionLevels = new IntArrayList();
+            fieldRepetitionLevels = new IntArrayList();
+            NamedTypeSignature namedTypeSignature = fields.get(i).getNamedTypeSignature();
             Type fieldType = typeManager.getType(namedTypeSignature.getTypeSignature());
             String name = namedTypeSignature.getName();
-            blocks[i] = readBlock(name, fieldType, path, new IntArrayList());
+            blocks[i] = readBlock(name, fieldType, path, fieldDefinitionLevels, fieldRepetitionLevels);
+        }
+        BooleanList structIsNull = new BooleanArrayList();
+        IntList structOffsets = new IntArrayList();
+        calculateStructOffsets(toArray(path), structOffsets, structIsNull, definitionLevels, repetitionLevels, fieldDefinitionLevels, fieldRepetitionLevels);
+        return new RowBlock(0, structIsNull.size(), structIsNull.toBooleanArray(), structOffsets.toIntArray(), blocks);
+    }
+
+    /**
+     * Each struct has three variants of presence:
+     * 1) Struct is not defined, because one of it's optional parent fields is null
+     * 2) Struct is null
+     * 3) Struct is defined and not empty. In this case offset value is increased by one.
+     * <p>
+     * One of the struct's field repetition/definition levels are used to calculate offsets of the struct.
+     */
+    private void calculateStructOffsets(String[] path, IntList structOffsets, BooleanList structIsNull, IntList definitionLevels, IntList repetitionLevels, IntList fieldDefinitionLevels, IntList fieldRepetitionLevels)
+    {
+        int offset = 0;
+        structOffsets.add(offset);
+        if (fieldDefinitionLevels != null) {
+            int maxDefinitionLevel = getMaxDefinitionLevel(path);
+            int maxRepetitionLevel = getMaxRepetitionLevel(path);
+            for (int i = 0; i < fieldDefinitionLevels.size(); i++) {
+                if (fieldRepetitionLevels.get(i) <= maxRepetitionLevel) {
+                    if (isNullField(isRequired(path), fieldDefinitionLevels.get(i), maxDefinitionLevel)) {
+                        structIsNull.add(true);
+                        structOffsets.add(offset);
+                    }
+                    else if (isStructDefined(fieldDefinitionLevels, maxDefinitionLevel, i)) {
+                        structIsNull.add(false);
+                        offset++;
+                        structOffsets.add(offset);
+                    }
+                }
+            }
+            definitionLevels.addAll(fieldDefinitionLevels);
+            repetitionLevels.addAll(fieldRepetitionLevels);
         }
+    }
+
+    private boolean isStructDefined(IntList fieldDefinitionLevels, int maxDefinitionLevel, int i)
```

Comment:
please rename `i` to something more descriptive.

Reply:
you can inline this method above as it's only used from `calculateStructOffsets` and add a comment that says you are checking whether struct is defined.